doc/design/custom_xcode_schemes.md

```diff
@@ -0,0 +1,467 @@
+#   //Sources/App = ios_application
+#   //Sources/AppUITests = ios_ui_test
+
+load(
+    "@com_github_buildbuddy_io_rules_xcodeproj//xcodeproj:xcodeproj.bzl",
+    "xcode_schemes",
+    "xcodeproj",
+)
+
+_SCHEMES = [
+    xcode_schemes.scheme(
+        name = "Foo Module",
+        build_action = xcode_schemes.build_action(["//Sources/Foo"]),
+        test_action = xcode_schemes.test_action(["//Tests/FooTests"]),
+    ),
+    xcode_schemes.scheme(
+        name = "My Application",
+        test_action = xcode_schemes.test_action([
+            "//Tests/AppUITests",
+            "//Tests/FooTests",
+        ]),
+        launch_action = xcode_schemes.launch_action("//Sources/App"),
+    ),
+]
+
+xcodeproj(
+    name = "generate_xcodeproj",
+    project_name = "Command Line",
+    tags = ["manual"],
+    schemes = _SCHEMES,
+)
+```
+
+The above example adds a second scheme called `My Application`. It is configured with two test targets
+and a single launch target. All three targets are implicitly included in the build list.
+
+### Specifying Launch Arguments and Environment Variables
+
+There are times when it is desirabe to customize the launch action with arguments and environment
+variables. This can be accomplished by specifying them as parameters to the
+`xcode_schemes.launch_action` function call.
+
+```python
+# Assumptions
+#   //Sources/Foo:Foo - swift_library
+#   //Sources/FooTests:FooTests = ios_unit_test
+#   //Sources/App = ios_application
+#   //Sources/AppUITests = ios_ui_test
+
+load(
+    "@com_github_buildbuddy_io_rules_xcodeproj//xcodeproj:xcodeproj.bzl",
+    "xcode_schemes",
+    "xcodeproj",
+)
+
+_SCHEMES = [
+    xcode_schemes.scheme(
+        name = "Foo Module",
+        build_action = xcode_schemes.build_action(["//Sources/Foo"]),
+        test_action = xcode_schemes.test_action(["//Tests/FooTests"]),
+    ),
+    xcode_schemes.scheme(
+        name = "My Application",
+        test_action = xcode_schemes.test_action([
+            "//Tests/AppUITests",
+            "//Tests/FooTests",
+        ]),
+        launch_action = xcode_schemes.launch_action(
+            target = "//Sources/App",
+            args = [
+                "--my_awesome_flag",
+                "path/to/a/file.txt",
+            ],
+            env = {
+                "RELEASE_THE_KRAKEN": "true",
+            },
+        ),
+    ),
+]
+
+xcodeproj(
+    name = "generate_xcodeproj",
+    project_name = "Command Line",
+    tags = ["manual"],
+    schemes = _SCHEMES,
+)
+```
+
+This example is the same as the previous one except arguments and environment variables are
+defined for the launch action.
+
+## Introduction of Scheme Autogeneration Mode
+
+With this proposal, it is now possible to control how Xcode scheme autogeneration works using
+the `scheme_autogeneration_mode` attribute on `_xcodeproj`. 
+
+__Values__
+- `auto`: If no custom schemes are provided, an Xcode scheme will be created for every buildable
```

Comment:
If there are no custom schemes provided and the mode is `auto`, why wouldn't schemes be created just for the list of `targets` explicitly specified? To get schemes for all buildable transitive deps you'd need to specify `all`.

Reply:
Sorry. My wording is imprecise. The intention is for `auto` to be work like `all` when no custom schemes exist. I was trying to preserve the current behavior.

Reply:
That part was actually clear to me, I'm suggesting an alternative which would allow reducing the scheme pollution by default when using `targets` instead of `schemes` while still having a way to generate schemes for all transitive deps using `all`.

Reply:
I could see that being another option, where it's `top_level_only`.

Reply:
At Lyft, our generated Xcode project can be configured to go up and/or down N levels following the dependency graph for either transitive or reverse dependencies.

Reply:
But that's Focused Projects, not scheme control, right? Focused Projects will only have schemes for targets that are in-focus, so the behavior would be the same then.

So if you had Focused Projects, would you use need this different generation mode?

Reply:
I was under the impression that a "focused project" meant that there would only be schemes only for a subset of the build graph. Are those not the same thing?

Reply:
Focused Project means that there will only be targets and their associated files in Xcode. (It's not just schemes that slow down Xcode, the targets themselves do as well.) Schemes can only be for targets that are in Xcode, so if you have a Focused Project it will have fewer schemes. But you can have a non-Focused Project (maybe you want all the targets and files) and still want fewer schemes, which is where I thought `top_level_only` might be useful.

Reply:
ok, so if I'm understanding correctly, limiting the created schemes to a subset of the build graph doesn't mean the project is focused.

Only limiting the source files or targets to a subset of the build graph is referred to as a focused project.

Reply:
Correct. We will need a glossary soon 😅.

Reply:
I really like the addition of `top_level_only`. In our project we have schemes for the top-level buildables (such as apps and extensions mostly). These would cover the majority of the use-cases. Via the separate API it would then be possible to specify additional schemes for other library and test targets.

Reply:
So, to close this out, do we want the `top_level_only` value? Will you use it day one?

Reply:
We want it, and it seems @BalestraPatrick would use it regardless of Focused Projects support.

Reply:
I'd use `top_level_only` even without custom schemes if it's available before because I do see a perf hit in the scheme selector that is improved when trimming the schemes.

Reply:
I will add it.